src/webgpu/util/math.ts

```diff
@@ -875,3 +875,35 @@ export function cartesianProduct<T>(...inputs: T[][]): T[][] {
 
   return result;
 }
+
+/** @returns all of the permutations of an array
+ *
+ * Recursively calculates all of the permutations, does not cull duplicate
```

Comment:
`does not cull duplicate entries` - what do you mean by this? Seems like an unexpected thing for this function to do.

Reply:
If you were ask for permutations of [1, 1] it will return [[1, 1], [1, 1]], instead of just [1,1]. The caller should remove duplicates in the return if they care.

Reply:
culling duplicates is a pain, because we are operating on array of arrays, so .includes() doesn't work, since it will do a shallow comparison of the objects, not an element-by-element comparison of the contents.